12-offer-encoding.md

```diff
@@ -0,0 +1,823 @@
+    1. type: 12 (`features`)
+    2. data:
+        * [`...*byte`:`features`]
+    1. type: 14 (`absolute_expiry`)
+    2. data:
+        * [`tu64`:`seconds_from_epoch`]
+    1. type: 16 (`paths`)
+    2. data:
+        * [`...*blinded_path`:`paths`]
+    1. type: 20 (`issuer`)
+    2. data:
+        * [`...*utf8`:`issuer`]
+    1. type: 22 (`quantity_min`)
+    2. data:
+        * [`tu64`:`min`]
+    1. type: 24 (`quantity_max`)
+    2. data:
+        * [`tu64`:`max`]
+    1. type: 30 (`node_id`)
+    2. data:
+        * [`point32`:`node_id`]
+    1. type: 54 (`send_invoice`)
+    1. type: 34 (`refund_for`)
+    2. data:
+        * [`sha256`:`refunded_payment_hash`]
+    1. type: 240 (`signature`)
+    2. data:
+        * [`bip340sig`:`sig`]
+
+1. subtype: `blinded_path`
+2. data:
+   * [`point`:`blinding`]
+   * [`byte`:`num_hops`]
+   * [`num_hops*onionmsg_path`:`path`]
+
+## Requirements For Offers
+
+A writer of an offer:
+  - if it sets `node_id`:
+    - MUST set `node_id` to the public key of the node to request the invoice from.
+  - otherwise:
+    - MUST provide at least one `blinded_path`
+    - MUST use the final `onionmsg_path` `point` in the first `blinded_path` as the implied `node_id` for `signature`.
+  - MAY specify exactly one signature TLV: `signature`:
+    - If so, it MUST set `sig` to the signature using `node_id` as described in [Signature Calculation](#signature-calculation).
+  - MUST set `description` to a complete description of the purpose
+    of the payment.
+  - if the chain for the invoice is not solely bitcoin:
+    - MUST specify `chains` the offer is valid for.
+  - otherwise:
+    - MAY omit `chains`, implying that bitcoin is only chain.
+  - if a specific minimum `amount` is required for successful payment:
+    - MUST set `amount` to the amount expected (per item).
+    - if the currency for `amount` is that of the first entry in `chains`:
+      - MUST specify `amount` in multiples of the minimum lightning-payable unit
+        (e.g. milli-satoshis for bitcoin).
+    - otherwise:
+      - MUST specify `iso4217` as an ISO 4712 three-letter code.
+      - MUST specify `amount` in the currency unit adjusted by the ISO 4712
+        exponent (e.g. USD cents).
+  - if it supports offer features:
+    - SHOULD set `features` to the bitmap of offer features.
+  - if the offer expires:
+    - MUST set `absolute_expiry` `seconds_from_epoch` to the number of seconds
+      after midnight 1 January 1970, UTC that invoice_request should not be
+      attempted.
+  - if it is connected only by private channels:
+    - MUST include `paths` containing one or more paths to the node from
+      publicly reachable nodes.
+  - otherwise:
+    - MAY include `paths`.
+  - if it includes `paths`:
+    - if the `node_id` is not public:
+      - MUST ignore any invoice_request which does not use one of the `paths`.
+    - otherwise:
+      - MAY ignore any invoice_request which does not use one of the `paths`.
```

Comment:
Could this be simplified by making `node_id` and `paths` mutually exclusive? What would the rationale be for setting both?

Reply:
I think paths should be mandatory (https://github.com/rustyrussell/lightning-rfc/pull/4) but regardless it's useful to keep `node_id` to have a stable identifier for the merchant (which could be the `node_id` of a public node or not). That's because when you build a blinded path you can't control the last blinded id.

Reply:
This was discussed in yesterday's spec meeting. I realized that additionally we may want to require a `node_id` if a signature is present. @rustyrussell noted that we may want to require `node_id` regardless, IIRC. @thomash-acinq support making `paths` required as noted in this discussion. @rustyrussell Thought's on making both required?